### PR TITLE
feat(j-s): FormStepper refactoring

### PR DESCRIPTION
**Pull request by oddsson**
# FormStepper refactoring

[Asana](https://app.asana.com/0/1199153462262248/1203449306021105/f)

## What

Refactor FormStepper functionality. This includes deprecating the existing `FormStepper` component in island-ui.

## Why

The existing FormStepper component in island-ui violates [the open-closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle). You could only use links in the old form stepper but we feel that limitation should not be set by the component (you should be able to render whatever component you want in the form stepper). It is the developer's responsibility implement the component according to the design.

In our case, we needed a button in the stepper instead of links to be able to run some code when using the FormStepper.

## Screenshots / Gifs

https://user-images.githubusercontent.com/3789875/207574126-21f93cc2-ba1f-4916-82df-7191153ff200.mov

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/island-ui/core/src/lib/FormStepper/SubSections/SubSections.tsx [hunk 1]
```diff
@@ -53,6 +55,7 @@ export const SubSections: FC<{
               }
               showIcon={showSubSectionIcon}
               href={subSection.href}
+              onClick={subSection.onClick}
```
Comment (oddsson) [inline libs/island-ui/core/src/lib/FormStepper/SubSections/SubSections.tsx:58]:
```suggestion
```

File: apps/judicial-system/web/src/routes/Court/InvestigationCase/HearingArrangements/HearingArrangements.tsx [hunk 2]
```diff
@@ -96,38 +97,45 @@ const HearingArrangements = () => {
...
-            ? formatDateForServer(new Date(courtDate))
-            : undefined,
-          force: true,
-        },
-      ],
+      if (hasSentNotification && !courtDateHasChanged) {
+        router.push(`${destination}/${workingCase.id}`)
+      } else {
+        setNextRoute(`${destination}/${workingCase.id}`)
+        setModalVisible(true)
```
Comment (Arnarkari93) [inline apps/judicial-system/web/src/routes/Court/InvestigationCase/HearingArrangements/HearingArrangements.tsx:123]: Consider refactoring the use of these `nextRoute` states to be a part of setting the modal to visible. That way you will never render a modal that expects a destination route without one. This applies to all use-cases of this `nextRoute` states.

File: apps/judicial-system/web/src/utils/formHelper.ts [hunk 3]
```diff
@@ -169,3 +171,145 @@ export const hasDateChanged = (
...
+export const stepValidations = (theCase: Case): stepValidationsType => {
+  return {
+    [constants.CREATE_RESTRICTION_CASE_ROUTE]: validations.isDefendantStepValidRC(
+      theCase,
+      theCase.policeCaseNumbers,
+    ),
+    [constants.CREATE_TRAVEL_BAN_ROUTE]: validations.isDefendantStepValidRC(
+      theCase,
+      theCase.policeCaseNumbers,
+    ),
```
Comment (Arnarkari93) [inline apps/judicial-system/web/src/utils/formHelper.ts:219]: We do not need to run all the validation functions until the `caller` of this function has decided which routes to validate.
```suggestion
export const stepValidations = (): stepValidationsType => {
  return {
    [constants.CREATE_RESTRICTION_CASE_ROUTE]: (theCase: Case) => validations.isDefendantStepValidRC(
      theCase,
      theCase.policeCaseNumbers,
    ),
    [constants.CREATE_TRAVEL_BAN_ROUTE]: (theCase: Case) validations.isDefendantStepValidRC(
      theCase,
      theCase.policeCaseNumbers,
    ),
```

File: apps/judicial-system/web/src/utils/formHelper.ts [hunk 4]
```diff
@@ -169,3 +171,145 @@ export const hasDateChanged = (
...
+  const validationEntries = Object.entries(validations)
+  const stepsToCheck = validationEntries.filter(([key]) => steps.includes(key))
+
+  if (stepsToCheck.every(([, value]) => value === true)) {
+    return steps[steps.length - 1]
+  }
+
+  const [key] = stepsToCheck.find(([, value]) => value === false) || []
+  return key
+}
```
Comment (Arnarkari93) [inline apps/judicial-system/web/src/utils/formHelper.ts:315]:
```suggestion
export const findFirstInvalidStep = (steps: string[], theCase: Case) => {
  const validations = stepValidations()
  const validationFunctions = Object.entries(validations)
  const stepsToCheck = validationEntries.filter(([key]) => steps.includes(key))

  if (stepsToCheck.every(([, validationFn]) => validationFn(theCase) === true)) {
    return steps[steps.length - 1]
  }

  const [key] = stepsToCheck.find(([, validationFn]) => validationFn(theCase) === false) || []
  return key
}
```

File: libs/island-ui/core/src/lib/FormStepper/SubSectionItem/SubSectionItem.tsx [hunk 5]
```diff
@@ -49,9 +53,20 @@ export const SubSectionItem: FC<SubSectionItemProps> = ({
         )}
       </SectionNumberColumn>
       {href && currentState !== 'active' ? (
-        <Link href={href} underline="small" underlineVisibility="hover">
+        <Link href={href} underline="small">
           {renderChildren()}
         </Link>
+      ) : onClick && currentState !== 'active' ? (
```
Comment (Arnarkari93) [inline libs/island-ui/core/src/lib/FormStepper/SubSectionItem/SubSectionItem.tsx:59]: Do we need these changes after introducing the V2?

**Review by Arnarkari93 — CHANGES_REQUESTED** [review]
Good work! I have some suggestions/nitpicks.

**Review by Arnarkari93 — APPROVED** [review]
🙌

**Pull request merged**